### ISSUE TITLE
BrokenProcessPool when using multiple workers

### ISSUE THREAD

**Issue by shz9** [pystatgen/bio2zarr#78]
I've run into this issue a couple of times while attempting to convert 1000G data `chr22` VCF to Zarr format. The first time I tried calling `vcf2zarr convert ...` directly. The second time I did it in the staggered approach suggested in the documentation. However, it still produced this error at the `encode` stage, particularly when encoding 2D matrices:

```bash
(bio2zarr_env) [szabad@narval2 bio2zarr_experiments]$ vcf2zarr encode -p 4 converted/chr22.exploded converted/chr22.zarr
Encode 1D: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████| 19.7k/19.7k [09:35<00:00, 34.2chunks/s]
Encode 2D:  79%|███████████████████████████████████████████████████████████████████████████████▎                     | 8.49k/10.8k [28:22<29:35, 1.30chunks/s]Traceback (most recent call last):
  File "/home/szabad/bio2zarr_env/bin/vcf2zarr", line 8, in <module>
    sys.exit(vcf2zarr())
  File "/home/szabad/bio2zarr_env/lib/python3.9/site-packages/click/core.py", line 1157, in __call__
    return self.main(*args, **kwargs)
  File "/home/szabad/bio2zarr_env/lib/python3.9/site-packages/click/core.py", line 1078, in main
    rv = self.invoke(ctx)
  File "/home/szabad/bio2zarr_env/lib/python3.9/site-packages/click/core.py", line 1688, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/szabad/bio2zarr_env/lib/python3.9/site-packages/click/core.py", line 1434, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/szabad/bio2zarr_env/lib/python3.9/site-packages/click/core.py", line 783, in invoke
    return __callback(*args, **kwargs)
  File "/home/szabad/bio2zarr_env/lib/python3.9/site-packages/bio2zarr/cli.py", line 127, in encode
    vcf.encode(
  File "/home/szabad/bio2zarr_env/lib/python3.9/site-packages/bio2zarr/vcf.py", line 1538, in encode
    SgvcfZarr.encode(
  File "/home/szabad/bio2zarr_env/lib/python3.9/site-packages/bio2zarr/vcf.py", line 1498, in encode
    pwm.submit(
  File "/home/szabad/bio2zarr_env/lib/python3.9/site-packages/bio2zarr/core.py", line 219, in __exit__
    wait_on_futures(self.futures)
  File "/home/szabad/bio2zarr_env/lib/python3.9/site-packages/bio2zarr/core.py", line 56, in wait_on_futures
    raise exception
concurrent.futures.process.BrokenProcessPool: A process in the process pool was terminated abruptly while the future was running or pending.
```

I'm running this on a cluster and it's possible that the processes are killed silently if they used more resources than they should. But in this case, I'm just using 4 workers and it didn't seem they were using lots of memory.

**Also a general question related to this**: Do you think it's possible to pick up the encoding work from where it left off if things like this happen instead of starting over?

**Comment by jeromekelleher** [comment]
I just hit the same issue and it was due to the worker getting killed by the OOM killer.

I suspect what happened here is that you had just enough memory reserved for 4 workers for all the fields except PL. These fields are huge (each chunk is nearly 1GB), so I'm not surprised the cluster killed it. 

This is not obvious, so we should potentially intercept the BrokenProcessPool exception in the main process and add an informative mesage like "you probably ran out of memory".

I think the simplest thing for now is to just remove PL from your experiments. Edit the schema JSON and delete the PL field, and it should all work fine.

> Also a general question related to this: Do you think it's possible to pick up the encoding work from where it left off if things like this happen instead of starting over?

I think that's closely related to how we're going to split this up into manageable bits for cluster scheduling. See #71 and #77 for discussions on how we're doing this for ``explode`` (and I think some high-level discussion about ``encode`` too).

**Comment by benjeffery** [comment]
I'm hitting this exact same exception - the encode keeps going which isn't ideal.

**Comment by jeromekelleher** [comment]
Good to know, I think I know how to fix

**Comment by jeromekelleher** [comment]
Ooh, ``max-memory`` is a great idea! We could associate a memory value with each future (say 3 times the number of bytes in one chunk of the array) and then stop submitting when the total for the outstanding futures exceeds this. I expect this would work quite well, especially if we try and mix up the big chunks with smaller ones.

We should follow this up in a separate issue

**Comment by jeromekelleher** [comment]
I pushed an update in #80 which should help debug this @shz9. If you run ``encode`` with ``-v`` it should give you helpful messages about the minimum RAM required per array.

Doing things better will require some refactoring, which we should probably do as part of making the encode job work in parallel over a cluster.

**Comment by jeromekelleher** [comment]
Closing this as we've added the max-memory argument as well.
